doc/case-studies/xycloans/xycloans_monitor.tla

```diff
@@ -0,0 +1,152 @@
+-------------------------- MODULE xycloans_monitor --------------------------
+(*
+ * A monitor for the xycLoans contract to detect the known issue.
+ *
+ * This is a manual translation of the Typescript monitor from
+ * verify_js_examples_xycloans.ts. Our goal is to produce a simple and readable
+ * TLA+ specification that is used to monitor existing transactions and
+ * generate new transactions, e.g., from the current state.
+ *
+ * After receiving feedback from the users, we would think about automation.
+ *
+ * Igor Konnov, 2024
+ *)
+
+EXTENDS Integers, xycloans_types
+
+STROOP == 10000000
+
+CONSTANT
+    \* The token address for the xycLoans contract.
+    \* @type: Str;
+    XLM_TOKEN_SAC_TESTNET
+
+(* The internal state of our monitor (not of the contract) *)
+VARIABLES
+    \* The last monitored transaction.
+    \* @type: $tx;
+    last_tx,
+    \* Shares per address.
+    \* @type: Str -> Int;
+    shares,
+    \* The sum over all shares.
+    \* @type: Int;
+    total_shares,
+    \* Fee per share for the entire pool, in stroops.
+    \* @type: Int;
+    fee_per_share_universal
+
+(* The core logic of the monitor for the contract data *)
+
+\* @type: ($tx, Bool) => Bool;
+reverts_if(tx, cond) == cond => ~tx.status
+\* @type: ($tx, Bool) => Bool;
+succeeds_with(tx, cond) == tx.status => cond
+\* @type: (Str -> a, Str, a) => a;
+get_or_else(map, key, default) ==
+    IF key \in DOMAIN map THEN map[key] ELSE default
+\* integer division with rounding up
+div_ceil(a, b) == (a + (b - 1)) \div b
+\* integer division with rounding down
+div_floor(a, b) == a \div b
+\* @type: ($env, Str) => Int;
+token_balance(env, a) == get_or_else(env.storage.token_persistent.Balance, a, 0)
+\* @type: ($env, Str) => Int;
+old_token_balance(env, a) == get_or_else(env.old_storage.token_persistent.Balance, a, 0)
+
+\* @type: $tx => Bool;
+initialize(tx) ==
+    LET call == AsInitialize(tx.call) IN
+    /\ IsInitialize(tx.call)
+    /\ reverts_if(tx, tx.env.old_storage.self_instance.TokenId /= "")
+    /\ succeeds_with(tx, tx.env.storage.self_instance.TokenId = XLM_TOKEN_SAC_TESTNET)
+    \* these conditions are not required by a monitor, but needed to avoid spurious generated values
+    /\ succeeds_with(tx,
+        tx.env.storage.self_instance.FeePerShareUniversal = tx.env.old_storage.self_instance.FeePerShareUniversal)
+    \* these conditions are not required by a monitor, but needed to avoid spurious generated values
+    /\ succeeds_with(tx,
+        tx.env.storage.self_persistent = tx.env.old_storage.self_persistent)
+    /\ last_tx' = tx
```

Comment:
Should we move the `last_tx` assignment into the `MC_` file?

I think it should happen at the same level as the `storage` assignment.

Reply:
Yep, totally